### PR TITLE
Fix for #21243 - Glassfish ignores variables in asenv.conf that begin with the letter "S"

### LINKED ISSUE #21243: Glassfish ignores variables in asenv.conf that begin with the letter "S"
#Any variables I set in asenv.conf that begin with a capital letter "S" are ignored and not accessible to web applications. All other variables are set and can be retrieved as expected. The problem may also happen with a lowercase letter "s", but I didn't specifically test that case.
#### Environment
Mac OS 10.9.5
Linux version 2.6.32-431.29.2.el6.x86_64 (mockbuild@c6b9.bsys.dev.centos.org) (gcc version 4.4.7 20120313 (Red Hat 4.4.7-4) (GCC) )  SMP Tue Sep 9 21:36:05 UTC 2014
#### Affected Versions
[4.1]

### PR DESCRIPTION
**Pull request by kokil12**
Fixes #21243 - The bug was in the regex pattern of function "parseAsEnv()" of file "MainHelper.java". The regex pattern has been changed and have been manually tested. The code gives the expected output with the suggested changes.

### CODE REVIEW

**Comment by kokil12** [comment]
@glassfishrobot Run CI tests please

**Comment by kokil12** [comment]
@glassfishrobot Run CI tests please

**Comment by bshannon** [comment]
The change won't match "Set A=B".  Shouldn't you use the "ignore case" option for the match?

**Comment by kokil12** [comment]
The regex pattern has been modified to support case insensitivity. Now the change is able to match Set A=B also.

**Comment by bshannon** [comment]
The change looks good now, but it seems like there could be a unit test to test the different cases that it intends to support:

set A=B
SET A=B
Set A=B
set A="B"
etc.

**Comment by kokil12** [comment]
Added unit test to test the regex pattern for different cases.

File: main/nucleus/core/bootstrap/src/test/java/com/sun/enterprise/glassfish/bootstrap/MainHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.sun.enterprise.glassfish.bootstrap.MainHelper;
+import java.io.*;
+import java.util.*;
+
+public class MainHelperTest {
+    /* This test is used to test the regex pattern of "parseAsEnv" method of "MainHelper.java"*/
+    @Test
+    public void parseAsEnvTest() {
+        try {
+            Properties asenvProps = new Properties();
```
Comment (bshannon) [inline main/nucleus/core/bootstrap/src/test/java/com/sun/enterprise/glassfish/bootstrap/MainHelperTest.java]: Why initialize it with a Properties object if parseAsEnv is going to return you a new Properties object?

File: main/nucleus/core/bootstrap/src/test/java/com/sun/enterprise/glassfish/bootstrap/MainHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            String currentDir = System.getProperty("user.dir");
```
Comment (bshannon) [inline main/nucleus/core/bootstrap/src/test/java/com/sun/enterprise/glassfish/bootstrap/MainHelperTest.java]: This seems to depend on the test files being in a certain directory that corresponds to the user.dir System property, but it's not clear to me how that's set up or what guarantees that.  Wouldn't it be better to create a temporary directory, create a file there, write the test cases into the file, and then pass the temporary directory to the parseAsEnv method?  Or, if there's some other guarantee that user.dir wil be set properly, document how that works.

**Comment by kokil12** [comment]
Briefly in recent changes, "asenv.conf" and "asenv.bat" has been moved under " main/nucleus/core/bootstrap/src/test/resources/config/" . 
File "pom.xml" has been edited to include "maven-resources-plugin" which will copy the test resources residing under "/src/test/resources" to the temp directory "/tmp/test" . The copying takes place during maven "validate" phase.
So, now we are passing the path of the tempDir (/tmp/test) to the "parseAsEnv()" method of "MainHelper.java".

**Comment by bshannon** [comment]
Please add a comment to the unit test explaining this.

Also, it **really** would be better if the name of the test directory were generated so as to be unique and not conflict with other processes running on the same machine.

**Comment by kokil12** [comment]
Comment has been added to the unit test and the name of the test directory has been changed.

**Comment by bshannon** [comment]
Thanks for adding the comment, although I would prefer that it stay within 80 characters so that I don't need to use my entire screen width to read it.

Changing the name of the directory doesn't prevent conflicts if two instances of the same test are running on the same machine.  Here's what I did for a similar case in a TCK test:

```
            File home = File.createTempFile("javahome", "conf");
            home.delete();      // delete the temp file
            home.mkdir();       // reuse the name for a directory
            home.deleteOnExit();
            File conf = new File(home, "conf");
            conf.mkdir();
            conf.deleteOnExit();
            File mimetypes = new File(conf, "mime.types");
            mimetypes.deleteOnExit();
            PrintWriter pw = new PrintWriter(mimetypes);
            pw.println(entry1);
            pw.close();
```

It creates a temporary file (which will have a unique name), recreates it as a directory, creates another directory under it (needed for this test), then creates the test file in the directory.  The top level directory is then passed to the test method.

**Review by bshannon — DISMISSED** [review]
I'm going to approve this because some test is better than no test, although I would still prefer that the test was improved as I suggested.

**Comment by kokil12** [comment]
The test has been improved as suggested. Please verify.

**Review by bshannon — APPROVED** [review]
Looks great, thanks!

**Comment by arindam-bandyopadhyay** [comment]
@glassfishrobot Run CI tests please

**Comment by kokil12** [comment]
@glassfishrobot Run CI tests please

**Pull request merged**